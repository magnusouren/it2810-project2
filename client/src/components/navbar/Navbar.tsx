import { AppBar, IconButton, Toolbar, Typography } from '@mui/material';
import React from 'react';
import { Link } from 'react-router-dom';

import MovieIcon from '../../assets/SVG/camera-white.svg';
import Search from '../search/Search';
import LoginButton from './LoginButton';
import styles from './Navbar.module.scss';

/**
 *
 * Navbar component.
 *
 * @returns {React.JSX.Element}
 *
 */

const Navbar: React.FC = () => {
  return (
    <AppBar position='static'>
      <Toolbar className={styles.navbarToolbar}>
        <Link to='/' className={styles.navLink} data-testid='home-link'>
          <IconButton edge='start' color='inherit' aria-label='home'>
            <img src={MovieIcon} width='40' alt='yeha' data-testid='HomeIcon' />
            {/* <HomeIcon /> From MUI */}
            <Typography variant='h6' className={styles.navbarTitle} />
          </IconButton>
        </Link>
        <Search />
        <LoginButton />
      </Toolbar>
    </AppBar>
  );
};

export default Navbar;
