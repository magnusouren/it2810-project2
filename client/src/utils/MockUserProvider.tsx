import { FC, ReactNode, useState } from 'react';
import { vi } from 'vitest';

// Importer UserContext
import { UserContext } from '../context/UserContext'; // Importer UserContext
import { User } from '../types';

interface MockUserProviderProps {
  children: ReactNode;
  mockUser?: User;
  loginMock?: () => void;
  logoutMock?: () => void;
  deleteUserMock?: () => void;
  toggleDarkModeMock?: () => void;
}

const MockUserProvider: FC<MockUserProviderProps> = ({
  children,
  mockUser,
  loginMock = vi.fn(),
  logoutMock = vi.fn(),
  deleteUserMock = vi.fn(),
  toggleDarkModeMock = vi.fn(),
}) => {
  const [user, setUser] = useState<User | undefined>(mockUser);

  const login = loginMock;
  const logout = logoutMock;
  const deleteUser = deleteUserMock;
  const toggleDarkMode = toggleDarkModeMock;

  const value = {
    user,
    setUser,
    login,
    logout,
    deleteUser,
    darkMode: false,
    toggleDarkMode,
  };

  return <UserContext.Provider value={value}>{children}</UserContext.Provider>;
};

export default MockUserProvider;
