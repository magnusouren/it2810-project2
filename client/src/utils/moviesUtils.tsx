import { GET_MOVIES, GET_MOVIES_BY_GENRE, GET_MOVIES_BY_RATING, GET_MOVIES_BY_TITLE_AZ } from '../graphql/queries';

export const determineQueryAndVariables = (page: number, genre: string, sort: string, userId: string) => {
  const baseVariables = { page: page, userId: userId };
  const alphabeticalSort = sort === 'a-z' || sort === 'z-a';
  const ratingSort = sort === 'h-l' || sort === 'l-h';

  if (genre) {
    if (alphabeticalSort) {
      return {
        query: GET_MOVIES_BY_TITLE_AZ,
        variables: { ...baseVariables, genreId: genre, order: sort },
      };
    }
    if (ratingSort) {
      return {
        query: GET_MOVIES_BY_RATING,
        variables: { ...baseVariables, genreId: genre, order: sort },
      };
    }
    return {
      query: GET_MOVIES_BY_GENRE,
      variables: { ...baseVariables, genreId: genre },
    };
  }

  if (alphabeticalSort) {
    return {
      query: GET_MOVIES_BY_TITLE_AZ,
      variables: { ...baseVariables, order: sort },
    };
  }
  if (ratingSort) {
    return {
      query: GET_MOVIES_BY_RATING,
      variables: { ...baseVariables, order: sort },
    };
  }
  return {
    query: GET_MOVIES,
    variables: { ...baseVariables },
  };
};
