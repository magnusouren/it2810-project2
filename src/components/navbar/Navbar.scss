.navbar-toolbar {
    display: flex;
    justify-content: space-between;
  }
  
  .navbar-title {
    flex-grow: 1;
  }
  