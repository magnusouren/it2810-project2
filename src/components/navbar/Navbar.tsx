import { Home as HomeIcon } from '@mui/icons-material';
import { AppBar, IconButton, Toolbar, Typography } from '@mui/material';
import React from 'react';
import { Link } from 'react-router-dom';

import Search from '../search/Search';
import LoginButton from './LoginButton';
import styles from './Navbar.module.scss';

const Navbar: React.FC = () => {
  return (
    <AppBar position='static'>
      <Toolbar className={styles.navbarToolbar}>
        <Link to='/' className={styles.navLink}>
          <IconButton edge='start' color='inherit' aria-label='home'>
            <HomeIcon />
            <Typography variant='h6' className={styles.navbarTitle} />
          </IconButton>
        </Link>
        <Search />
        <LoginButton />
      </Toolbar>
    </AppBar>
  );
};

export default Navbar;
