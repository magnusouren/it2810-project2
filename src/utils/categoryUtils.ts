import categoryData from '../../storage/categories.json';
import { Category } from '../types';

export const getCategoryById = (id: number): Category | undefined => {
  const categories: Category[] = categoryData;

  return categories.find((category) => category.id === id);
};

export const getCategoryNameById = (id: number): string | undefined => {
  const category = getCategoryById(id);
  return category?.name;
};

export const getCategoryIds = (): number[] => {
  return categoryData.map((category) => category.id);
};

export const getCategories = (): Category[] => {
  return categoryData;
};

export const getCategoryNames = (): string[] => {
  return categoryData.map((category) => category.name);
};

export const getCategoryIdByName = (name: string): number | undefined => {
  const category = categoryData.find((category) => category.name === name);
  return category?.id;
};
