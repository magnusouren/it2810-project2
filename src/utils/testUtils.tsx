import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { UserProvider } from '../context/UserContext';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouterAndUserContext = (component: JSX.Element) =>
  render(
    <UserProvider>
      <BrowserRouter>{component}</BrowserRouter>
    </UserProvider>,
  );

const queryClient = new QueryClient();

// Function to render a component wrapped inside a browserrouter and queryclient for testing purposes
export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );
