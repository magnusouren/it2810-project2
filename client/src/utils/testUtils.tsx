import { ApolloClient, ApolloProvider, HttpLink, InMemoryCache } from '@apollo/client';
import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import { User } from '../types';
import MockUserProvider from './MockUserProvider';

const link = new HttpLink({
  uri: 'http://localhost:4000', // Endepunktet til din GraphQL server
});

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: link,
});

type ProviderMocks = {
  loginMock?: () => void;
  logoutMock?: () => void;
  deleteUserMock?: () => void;
  toggleDarkModeMock?: () => void;
};

interface ProviderProps {
  child: React.ReactNode;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  mocks?: any[];
  mockUser?: User;
  providerMocks?: ProviderMocks;
}

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithProviders = ({ child, mockUser, mocks, providerMocks }: ProviderProps) =>
  render(
    <MockUserProvider mockUser={mockUser} {...providerMocks}>
      <ApolloProvider client={client}>
        <MockedProvider mocks={mocks} addTypename={false}>
          <BrowserRouter>{child}</BrowserRouter>
        </MockedProvider>
      </ApolloProvider>
    </MockUserProvider>,
  );
