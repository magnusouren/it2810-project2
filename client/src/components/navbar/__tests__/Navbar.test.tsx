import { act, screen } from '@testing-library/react';

import { renderWithProviders } from '../../../utils/testUtils';
import Navbar from '../Navbar';

describe('Navbar', () => {
  it('Should match snapshot', () => {
    const { container } = renderWithProviders({ child: <Navbar /> });
    expect(container).toMatchSnapshot();
  });

  it('Should show home icon', () => {
    renderWithProviders({ child: <Navbar /> });

    expect(screen.getByTestId('HomeIcon')).toBeDefined();
  });

  it('Should navigate to / (home) when home icon is clicked', async () => {
    renderWithProviders({ child: <Navbar /> });
    act(() => {
      screen.getByTestId('home-link').click();
    });
    expect(window.location.pathname).toBe('/');
  });
});
