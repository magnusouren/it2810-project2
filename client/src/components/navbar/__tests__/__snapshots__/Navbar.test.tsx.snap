// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > Should match snapshot 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular _navbarToolbar_c9a110 css-hyum1k-MuiToolbar-root"
    >
      <a
        class="_navLink_c9a110"
        data-testid="home-link"
        href="/"
      >
        <button
          aria-label="home-button"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1d6w9lk-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <img
            alt="Bingewatcher icon"
            data-testid="HomeIcon"
            src="/project2/assets/SVG/camera-white.svg"
            width="40"
          />
          <h5
            class="MuiTypography-root MuiTypography-h5 _navbarTitle_c9a110 css-ag7rrr-MuiTypography-root"
          >
            Bingewatcher
          </h5>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </a>
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
        data-testid="login-button"
        tabindex="0"
        type="button"
      >
        <p
          class="_loginText_67d6a9"
        >
          Login
        </p>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-jj92jx-MuiSvgIcon-root"
          data-testid="AccountCircleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 4c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6zm0 14c-2.03 0-4.43-.82-6.14-2.88C7.55 15.8 9.68 15 12 15s4.45.8 6.14 2.12C16.43 19.18 14.03 20 12 20z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </header>
</div>
`;
